zkevm-circuits/src/mpt_circuit/rlp_gadgets.rs

```diff
@@ -0,0 +1,770 @@
+use crate::{
+    _cb, circuit,
+    circuit_tools::{
+        cached_region::{CachedRegion, ChallengeSet},
+        cell_manager::Cell,
+        constraint_builder::{ConstraintBuilder, RLCable, RLCableValue},
+    },
+    matchw,
+    mpt_circuit::{
+        helpers::FIXED,
+        param::{RLP_LIST_LONG, RLP_LIST_SHORT, RLP_SHORT},
+        FixedTableTag,
+    },
+    util::Expr,
+};
+use eth_types::Field;
+use gadgets::util::{not, pow, Scalar};
+use halo2_proofs::plonk::{Error, Expression};
+
+use super::{
+    helpers::MPTConstraintBuilder,
+    param::{KEY_PREFIX_ODD, KEY_TERMINAL_PREFIX_ODD, RLP_LONG},
+};
+
+// Decodes the first byte of an RLP data stream to return (is_list, is_short, is_long, is_very_long)
+pub(crate) fn decode_rlp(byte: u8) -> (bool, bool, bool, bool) {
+    if byte < RLP_LIST_SHORT {
+        const RLP_SHORT_INCLUSIVE: u8 = RLP_SHORT - 1;
+        const RLP_LONG_EXCLUSIVE: u8 = RLP_LONG + 1;
+        const RLP_VALUE_MAX: u8 = RLP_LIST_SHORT - 1;
+
+        let mut is_short = false;
+        let mut is_long = false;
+        let mut is_very_long = false;
+        match byte {
+            0..=RLP_SHORT_INCLUSIVE => is_short = true,
+            RLP_SHORT..=RLP_LONG => is_long = true,
+            RLP_LONG_EXCLUSIVE..=RLP_VALUE_MAX => is_very_long = true,
+            _ => unreachable!(),
+        }
+        (false, is_short, is_long, is_very_long)
+    } else {
+        const RLP_LIST_LONG_1: u8 = RLP_LIST_LONG + 1;
+        const RLP_LIST_LONG_2: u8 = RLP_LIST_LONG + 2;
+
+        let mut is_short = false;
+        let mut is_long = false;
+        let mut is_very_long = false;
+        match byte {
+            RLP_LIST_SHORT..=RLP_LIST_LONG => is_short = true,
+            RLP_LIST_LONG_1 => is_long = true,
+            RLP_LIST_LONG_2 => is_very_long = true,
+            _ => (),
+        }
+        (true, is_short, is_long, is_very_long)
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+pub(crate) struct RLPListGadget<F> {
+    pub(crate) is_short: Cell<F>,
+    pub(crate) is_long: Cell<F>,
+    pub(crate) is_very_long: Cell<F>,
+    pub(crate) is_string: Cell<F>,
+    pub(crate) bytes: Vec<Expression<F>>,
+}
+
+#[derive(Clone, Debug, Default)]
+pub(crate) struct RLPListWitness {
+    pub(crate) is_short: bool,
+    pub(crate) is_long: bool,
+    pub(crate) is_very_long: bool,
+    pub(crate) is_string: bool,
+    pub(crate) bytes: Vec<u8>,
+}
+
+impl<F: Field> RLPListGadget<F> {
```

Comment:
If I am not mistaken, there are no constraints that would ensure that the byte streams are properly RLP encoded currently. So, for example for a storage leaf
`
[226 160 59 138 106 70 105 186 37 13 38 205 122 69 158 202 157 33 95 131 7 227 58 235 229 3 121 188 90 54 23 236 52 68 1]
`
the RLP gadget decodes `226` to `is_list=true` and `is_short=true`, from which it then knows how to compute for example the number of the RLP bytes. But the byte stream like
`
[227 160 59 138 106 70 105 186 37 13 38 205 122 69 158 202 157 33 95 131 7 227 58 235 229 3 121 188 90 54 23 236 52 68 1 x]
`
would be accepted as well (as far as RLP gadget is concerned), although you would still need to find the hash collision to make the attack work.

In the old code I had constraints like
`
226 - 192 = 160 - 128 + 2
`
But indeed because of the hashing these might not be necessary.

@Brechtpd  Do you think that checking that the byte streams are properly RLP encoded is not necessary?

Reply:
Ah, yes, they are there! Sorry, my brain was formatted too much during vacation :)